cli/command/init/initiator.go

```diff
@@ -0,0 +1,445 @@
+package init
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/AlecAivazis/survey/v2/terminal"
+
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/saucelabs/saucectl/internal/config"
+	"github.com/saucelabs/saucectl/internal/credentials"
+	"github.com/saucelabs/saucectl/internal/devices"
+	"github.com/saucelabs/saucectl/internal/framework"
+	"github.com/saucelabs/saucectl/internal/rdc"
+	"github.com/saucelabs/saucectl/internal/region"
+	"github.com/saucelabs/saucectl/internal/resto"
+	"github.com/saucelabs/saucectl/internal/testcomposer"
+	"github.com/saucelabs/saucectl/internal/vmd"
+)
+
+var androidDevicesPatterns = []string{
+	"Amazon Kindle Fire .*", "Google Pixel .*", "HTC .*", "Huawei .*",
+	"LG .*", "Motorola .*", "OnePlus .*", "Samsung .*", "Sony .*",
+}
+
+var iOSDevicesPatterns = []string{"iPad .*", "iPhone .*"}
+
+type initiator struct {
+	stdio        terminal.Stdio
+	infoReader   framework.MetadataService
+	deviceReader devices.Reader
+	vmdReader    vmd.Reader
+
+	frameworks        []string
+	frameworkMetadata []framework.Metadata
+}
+
+// newInitiator creates a new initiator instance.
+func newInitiator(stdio terminal.Stdio, creds credentials.Credentials, regio string) *initiator {
+	r := region.FromString(regio)
+	tc := testcomposer.Client{
+		HTTPClient:  &http.Client{Timeout: testComposerTimeout},
+		URL:         r.APIBaseURL(),
+		Credentials: creds,
+	}
+
+	rc := rdc.Client{
+		HTTPClient: &http.Client{Timeout: rdcTimeout},
+		URL:        r.APIBaseURL(),
+		Username:   creds.Username,
+		AccessKey:  creds.AccessKey,
+	}
+
+	rs := resto.Client{
+		HTTPClient: &http.Client{Timeout: restoTimeout},
+		URL:        r.APIBaseURL(),
+		Username:   creds.Username,
+		AccessKey:  creds.AccessKey,
+	}
+
+	return &initiator{
+		stdio:        stdio,
+		infoReader:   &tc,
+		deviceReader: &rc,
+		vmdReader:    &rs,
+	}
+}
+
+func isNativeFramework(framework string) bool {
+	return framework == config.KindEspresso || framework == config.KindXcuitest
+}
+
+func needsApps(framework string) bool {
+	return isNativeFramework(framework)
+}
+
+func needsCypressJSON(framework string) bool {
+	return framework == config.KindCypress
+}
+
+func needsDevice(framework string) bool {
+	return isNativeFramework(framework)
+}
+
+func needsEmulator(framework string) bool {
+	return framework == config.KindEspresso
+}
+
+func needsPlatform(framework string) bool {
+	return !isNativeFramework(framework)
+}
+
+func needsRootDir(framework string) bool {
+	return !isNativeFramework(framework)
+}
+
+func needsVersion(framework string) bool {
+	return !isNativeFramework(framework)
+}
+
+func (ini *initiator) configure() (*initConfig, error) {
```

Comment:
You could make this method less procedural if you had a subset of framework specific initializers that can return an `initConfig`.

Reply:
This has been splitted !